lib/MonadicRewrite.thy

```diff
@@ -255,29 +326,117 @@ lemma monadic_rewrite_symb_exec_l'_preserve_names:
      \<not> F \<longrightarrow> no_fail P' m;
      \<And>rv. monadic_rewrite F E (Q rv) (x rv) y;
      \<lbrace>P\<rbrace> m \<lbrace>Q\<rbrace> \<rbrakk>
-      \<Longrightarrow> monadic_rewrite F E (P and P') (m >>= (\<lambda>rv. x rv)) y"
+   \<Longrightarrow> monadic_rewrite F E (P and P') (m >>= (\<lambda>rv. x rv)) y"
   by (rule monadic_rewrite_symb_exec_l')
 
 (* FIXME merge into below upon change-over desribed above *)
 lemmas monadic_rewrite_symb_exec_l'_TT
-    = monadic_rewrite_symb_exec_l'_preserve_names[where P'="\<top>" and F=True, simplified]
+  = monadic_rewrite_symb_exec_l'_preserve_names[where P'="\<top>" and F=True, simplified]
 
 lemmas monadic_rewrite_symb_exec_l
-    = monadic_rewrite_symb_exec_l''[where F=True and P'=\<top>, simplified]
-      monadic_rewrite_symb_exec_l''[where F=False, simplified simp_thms]
+  = monadic_rewrite_symb_exec_l''[where F=True and P'=\<top>, simplified]
+    monadic_rewrite_symb_exec_l''[where F=False, simplified simp_thms]
+
+lemma monadic_rewrite_symb_exec_l_known:
+  "\<lbrakk> \<And>s. \<lbrace>(=) s\<rbrace> m \<lbrace>\<lambda>r. (=) s\<rbrace>; empty_fail m;
```

Comment:
Invariant notation?

Reply:
Spurred on by your suggestions, I went over the file looking for inv notation. Next push these should be resolved.